lib/registry.js

```diff
@@ -29,48 +34,99 @@ export default class Registry {
   // Public method
   create(): Provider {
     const provider = new Provider()
-    provider.onDidUpdate(() => {
-      this.emitter.emit('did-update')
+    provider.onDidAdd((status) => {
+      this.statusAdd(provider, status)
+    })
+    provider.onDidRemove((title) => {
+      this.statusRemove(provider, title)
     })
-    provider.onDidDestroy(() => {
-      this.subscriptions.remove(provider)
+    provider.onDidClear(() => {
+      this.statusClear(provider)
+    })
+    provider.onDidDispose(() => {
+      this.statusClear(provider)
       this.providers.delete(provider)
-      this.emitter.emit('did-update')
     })
-    this.subscriptions.add(provider)
     this.providers.add(provider)
     return provider
   }
-  getTitles(): Array<string> {
-    let texts = []
-    for (const provider of this.providers) {
-      texts = texts.concat(Array.from(provider.texts))
+  statusAdd(provider: Provider, status: { title: string, priority: number }): void {
+    for (let i = 0; i < this.itemsActive.length; i++) {
+      const entry = this.itemsActive[i]
+      if (entry.title === status.title && entry.provider === provider) {
+        // Item already exists, ignore
+        break
+      }
     }
-    return texts.sort(function(a, b) {
-      return a.priority - b.priority
-    }).map(function(item) {
-      return item.title
+
+    this.itemsActive.push({
+      title: status.title,
+      priority: status.priority,
+      provider,
+      timeAdded: Date.now(),
```

Comment:
You can see an example of using this [here](https://github.com/AtomLinter/linter-stylelint/blob/8f7e9410c0fe6aa99cf6be21470f701b0c0e1afe/lib/index.js#L37-L51).